### ISSUE TITLE
Error reading from database (latest fresh build)

### ISSUE THREAD

**Issue by mcmonkey4eva** [mbdavid/LiteDB#399]
```
System.InvalidCastException: Unable to cast object of type 'LiteDB.DataPage' to type 'LiteDB.IndexPage'.
   at LiteDB.PageService.GetPage[T](UInt32 pageID) in C:\Github\LiteDB\LiteDB\Engine\Services\PageService.cs:line 46
   at LiteDB.IndexService.GetNode(PageAddress address) in C:\Github\LiteDB\LiteDB\Engine\Services\IndexService.cs:line 325
   at LiteDB.IndexService.Find(CollectionIndex index, BsonValue value, Boolean sibling, Int32 order) in C:\Github\LiteDB\LiteDB\Engine\Services\IndexService.cs:line 367
   at LiteDB.QueryEquals.<ExecuteIndex>d__2.MoveNext() in C:\Github\LiteDB\LiteDB\Engine\Query\QueryEquals.cs:line 18
   at LiteDB.LiteEngine.<Find>d__49.MoveNext() in C:\Github\LiteDB\LiteDB\Engine\Engine\Find.cs:line 36
   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext() in C:\Github\LiteDB\LiteDB\Database\Collections\Find.cs:line 32
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
```
and
```
System.InvalidCastException: Unable to cast object of type 'LiteDB.ExtendPage' to type 'LiteDB.DataPage'.
   at LiteDB.PageService.GetPage[T](UInt32 pageID) in C:\Github\LiteDB\LiteDB\Engine\Services\PageService.cs:line 46
   at LiteDB.LiteEngine.<Find>d__49.MoveNext() in C:\Github\LiteDB\LiteDB\Engine\Engine\Find.cs:line 44
   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext() in C:\Github\LiteDB\LiteDB\Database\Collections\Find.cs:line 32
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)

```
two similar errors

I'm reading from and writing to the database asynchronously, both errors are from read ops.

Using, specifically, `LiteCollection<BsonDocument>.FindById(...)`

EDIT: Got another unique error message from the same situation.
```
System.ArgumentException: Offset and length were out of bounds for the array or count is gre
ater than the number of elements from index to the end of the source collection.
   at System.Buffer.BlockCopy(Array src, Int32 srcOffset, Array dst, Int32 dstOffset, Int32 count)
   at LiteDB.ByteReader.ReadBytes(Int32 count) in C:\Github\LiteDB\LiteDB\Utils\ByteReader.cs:line 108
   at LiteDB.BsonReader.ReadString(ByteReader reader) in C:\Github\LiteDB\LiteDB\Document\Bson\BsonReader.cs:line 145
   at LiteDB.BsonReader.ReadElement(ByteReader reader, String& name) in C:\Github\LiteDB\LiteDB\Document\Bson\BsonReader.cs:line 73
   at LiteDB.BsonReader.ReadArray(ByteReader reader) in C:\Github\LiteDB\LiteDB\Document\Bson\BsonReader.cs:line 52
   at LiteDB.BsonReader.ReadElement(ByteReader reader, String& name) in C:\Github\LiteDB\LiteDB\Document\Bson\BsonReader.cs:line 81
   at LiteDB.BsonReader.ReadDocument(ByteReader reader) in C:\Github\LiteDB\LiteDB\Document\Bson\BsonReader.cs:line 31
   at LiteDB.BsonSerializer.Deserialize(Byte[] bson) in C:\Github\LiteDB\LiteDB\Document\Bson\BsonSerializer.cs:line 25
   at LiteDB.LiteEngine.<Find>d__49.MoveNext() in C:\Github\LiteDB\LiteDB\Engine\Engine\Find.cs:line 0
   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext() in C:\Github\LiteDB\LiteDB\Database\Collections\Find.cs:line 32
   at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
```

**Comment by mbdavid** [comment]
Hi @mcmonkey4eva, can you past an unit test about this? Are you using Net35 ou NetStandard? I review all code to avoid this error.... it's all about locking.

**Comment by mcmonkey4eva** [comment]
tbh I don't know what triggered this beyond rapid async usage. I don't know if I'd be able to replicate it easily outside my environment. And it doesn't happen consistently, only once per few hundred calls.


It's the Net35 build, running under a .NET 4.5 program.

If you want, for a unit test, I could throw together something that just throws data into the wind asynchronously until it breaks, but I imagine you know your system well enough to make something at least as quality as that if not significantly more so.

**Comment by mbdavid** [comment]
How are you calling async? Can you paste you code in this part?

**Comment by mcmonkey4eva** [comment]
It more or less boils down to tasks, IE `Task.Factory.StartNew( () => { ... } );`
which in some cases does upserting of BsonDocument's.
IE, `BsonDocument bd = new BsonDocument(); bd["_id"] = GetID(); /* other data */ collection.Upsert(bd);`
and in other cases does reading of data via FindById
IE, `BsonDocument bd = collection.FindById(GetID());`

The program in question is a bit too complex and twisted to bother giving exact code, but it's basically the above in rapid parallel.
My idea of simulating it feasibly would be something like:
```cs
for (int i = 0; i < 1000; i++) // Change 1000 to whatever value spams it enough.
{
     int ind = i % 50;
     Task.Factory.StartNew(() => { collection.FindById(BitConverter.GetBytes(ind)); });
     Task.Factory.StartNew(() => {
        BsonDocument bd = new BsonDocument();
        bd["_id"] = new BsonValue(BitConverter.GetBytes(ind));
        collection.upsert(bd);
    });
}
```
... tweak as needed.

**Comment by mbdavid** [comment]
I made this unit test

```
[TestMethod]
public void Thread_FindUpsert_Test()
{
    using (var file = new TempFile())
    using (var db = new LiteEngine(file.Filename))
    {
        var tasks = new List<Task>();

        for (var i = 0; i < 10000; i++) // Change 1000 to whatever value spams it enough.
        {
            int ind = i % 50;
            var t1 = Task.Factory.StartNew(() => { db.FindById("col", BitConverter.GetBytes(ind)); });
            var t2 = Task.Factory.StartNew(() =>
            {
                var doc = new BsonDocument { { "_id", BitConverter.GetBytes(ind) } };;
                db.Upsert("col", doc);
            });

            tasks.AddRange(new [] { t1, t2 });
        }

        Task.WaitAll(tasks.ToArray());
    }
}
```

and still working. In this case, all tasks share same "db" instance. I will try more samples

**Comment by mbdavid** [comment]
Ok, I got the error. If I use `LiteDatabase` insted `LiteEngine` error occurs. Could be my `LiteCollection.Find` method and auto index create.

**Comment by mbdavid** [comment]
Hi @mcmonkey4eva, afer several day I still dont know how fix this in a good way. It's not possible catch exceptions when using yield return.

https://gist.github.com/mbdavid/6c693fc28ededabc8683de2f4d1364c9

Does anybody know how fix this?

**Comment by mcmonkey4eva** [comment]
Perhaps directly constructing a temporary list and returning that rather than doing a yield return? At least for the specific case... may require duplication of code.

**Comment by mbdavid** [comment]
If I use

```
    return FindWrap(query).ToArray();
```

works as expected. But, doing this it's terrible. This list could be too much documents and will use lots of memory.

My idea, at this moment, are:

1) Remove IndexNotFound exception
2) Run `EnsureIndex` as `GetCollection<T>` are called to all unique indexes (except PK)
3) Run AutoIndex as queries run (inside LiteEngine) but only non-unique

But, step 2 are not "free" and needs always read collection page.

**Comment by mkosieradzki** [comment]
@mbdavid Please have a look at the snippet I have added in the first comment to your gist. It allows error handling and still allows full streaming without loading everything into the memory. Not sure only where try-catch should be: in current or in MoveNext but that's easy to figure out :).

**Comment by mbdavid** [comment]
@mkosieradzki, I got me an idea: pass an `Action` to inner Find and avoid exception running index there. This will remove IndexNotFound exception.... I got it's simple way.

**Comment by mcmonkey4eva** [comment]
This happens MUCH less often in the newest build, but in a quick test it happened exactly once over a minute or two of testing (as opposed to several per second as it used to):
```
System.InvalidCastException: Unable to cast object of type 'LiteDB.ExtendPage' to type 'LiteDB.
DataPage'.
   at LiteDB.PageService.GetPage[T](UInt32 pageID) in C:\Github\LiteDB\LiteDB\Engine\Services\PageService.cs:line 46
   at LiteDB.LiteEngine.<Find>d__49.MoveNext() in C:\Github\LiteDB\LiteDB\Engine\Engine\Find.cs:line 46
   at LiteDB.LiteCollection`1.<Find>d__17.MoveNext() in C:\Github\LiteDB\LiteDB\Database\Collections\Find.cs:line 24
   at System.Linq.Enumerable.SingleOrDefault[TSource](IEnumerable`1 source)
```

**Comment by mcmonkey4eva** [comment]
I'm thinking this should be very high priority:
ANY rapid usage (async or sync) triggers this in recent builds.
I had to downgrade to the very first build of LiteDB v3, (Literally at commit "`Nuget v3`") to prevent the error. No modification to my code would let it through (I tried using my original code that I used prior to upsert, with an added synclock!)

**Comment by mbdavid** [comment]
Hi @mcmonkey4eva, yes this is priority and I was working on it. But was not easy found this tricky missing lock. From this `Nuget v3` commit a lot of changes was made to support back multi process too (in this commit there is no multi instances support).
Now I run this test over 100 times using 2000 loop and no exception was throwed. Please try again with lastest version.

**Comment by mcmonkey4eva** [comment]
That did it!

The newest build no longer throws even a single error in my testing.

Amazing work! Thank you!

**Comment by debagger** [comment]
Hello!
I had this error any time when work with db. It's about 600000 documents. Used latest (3.0.1) build from nuget. Its my first time exirience with LiteDB. Looks like broken documents in db. 
Code like that:
```
Dim col = db.GetCollection(Of DetailsLink)("DetailsLinks")
For Each item In col.FindAll
            Console.WriteLine(item.Uri)
Next
```

throw exception after some iterations.

**Comment by mcmonkey4eva** [comment]
@debagger should probably open a new issue for that.

Be sure to include a copy of an exception output when you do!
